### PR TITLE
Eva/lane detection

### PR DESCRIPTION
**Pull request by evamo0508**
#### Hardware
`$roslaunch auotvalet_lane_detection lane_detector.launch`

The launch file will be running the code under `scripts/`: `autovalet_lane_detector_classical.py`

Will include code for lane detection on Sim in another MR.

### CODE REVIEW

File: autovalet_husky/rviz/config.rviz [hunk 1]
```diff
@@ -6,9 +6,9 @@ Panels:
       Expanded:
```
Comment (kob51) [inline autovalet_husky/rviz/config.rviz]: not pushing rviz configs anymore, right?

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        th    = cv2.cvtColor(th, cv2.COLOR_HLS2RGB)
+        th    = cv2.cvtColor(th, cv2.COLOR_RGB2GRAY)
+        th    = cv2.GaussianBlur(th, (5, 5), 0)
+        _, th = cv2.threshold(th, 60, 255, cv2.THRESH_BINARY)
+	th    = cv2.morphologyEx(th, cv2.MORPH_OPEN, (5,5))
+
+        # extract ROI
+        mask = np.zeros_like(th)
+        mask[self.ROI_UPPER_Y:, :self.ROI_RIGHT_X] = 1
+        roi = cv2.bitwise_and(th, th, mask=mask)
```
Comment (kob51) [inline autovalet_lane_detection/scripts/autovalet_lane_detector.py]: comment these blocks a little more? i'm finding it hard to keep track of what is what

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        max_len, scenario, X1, Y1, X2, Y2 = 0, 0, 0, 0, 0, 0
+        if lines is not None: # 1. new line
+            for line in lines:
+                x1, y1, x2, y2 = line[0]
+                line_len = np.sqrt((x2-x1)**2 + (y2-y1)**2)
+                if max_len < line_len:
+                    max_len = line_len
+                    X1, Y1, X2, Y2 = x1, y1, x2, y2
+            scenario = 1
+            bb = (X1, Y1, np.abs(X2-X1), np.abs(Y2-Y1))
```
Comment (kob51) [inline autovalet_lane_detection/scripts/autovalet_lane_detector.py]: `bounding_box`

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+            self.tracker = cv2.TrackerKCF_create()
```
Comment (kob51) [inline autovalet_lane_detection/scripts/autovalet_lane_detector.py:96]: why don't you create in the constructor?

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        mask = np.zeros_like(th)
+        mask[self.ROI_UPPER_Y:, :self.ROI_RIGHT_X] = 1
+        roi = cv2.bitwise_and(th, th, mask=mask)
+
+        # detect center line
+        minLineLength, maxLineGap = 130, 20
+        edges = cv2.Canny(roi, 100, 200, apertureSize=3)
+        lines = cv2.HoughLinesP(edges, 1, np.pi/180, 30, np.array([]), minLineLength, maxLineGap)
+
+        # possible scenarios: 1. new line 2. tracking
```
Comment (kob51) [inline autovalet_lane_detection/scripts/autovalet_lane_detector.py]: what about 0?

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+                Y1 = min(Y1, img.shape[0] - 1)
+
+
+        # coordinates and slope
+        if scenario == 0:
+            return np.array([])
+        if X1 == X2: # vertical line
+            coordinates = [[X1, i] for i in range(Y2, Y1+1)]
+        else:
+            slope = 1.0 * (Y2 - Y1) / (X1 - X2) # assuming slope being positive, might have bug
```
Comment (kob51) [inline autovalet_lane_detection/scripts/autovalet_lane_detector.py]: why not `(X2 - X1)`?

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+            else:
+                coordinates = [[int(X1+i/slope), Y1-i] for i in range(Y1-Y2)]
+
+        return np.array(coordinates)
+
+    def line2cloud(self, depth, coordinates):
+        # ref: Subbu's code
+        u, v = coordinates[:, 0], coordinates[:, 1]
+        x    = (u - self.cx) / self.fx
+        y    = (v - self.cy) / self.fy
```
Comment (kob51) [inline autovalet_lane_detection/scripts/autovalet_lane_detector.py:136]: write the original matrix equation in the comments plz

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+        lane_width = 3.5 # meters, US standard
+
+        center_x = center_line_cloud[:, 0]
+        center_y = center_line_cloud[:, 1]
+        center_z = center_line_cloud[:, 2]
+
+        # calculate the mean of the points
+        center_mean = center_line_cloud.mean(axis=0)
+
+        # Do an SVD on the mean-centered data.
```
Comment (kob51) [inline autovalet_lane_detection/scripts/autovalet_lane_detector.py:158]: explain a little more what the result of svd is

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+        u, d, vh = np.linalg.svd(center_line_cloud - center_mean)
+        orth_vec = np.sign(vh[0][2]) * np.array([vh[0][2], -vh[0][0]])
+        #orth_vec = np.sign(vh[-1][0]) * np.array([vh[-1][0], vh[-1][2]])
+        norm_vec = orth_vec / np.linalg.norm(orth_vec)
+
+        # extend the points with lane_width
```
Comment (kob51) [inline autovalet_lane_detection/scripts/autovalet_lane_detector.py:167]: what frame's x,y,z? mention that here

File: autovalet_lane_detection/utils/hsv_check.py [hunk 10]
```diff
@@ -16,7 +16,7 @@ def nothing(x):
 
 while(1):
 
-    frame = cv2.imread('../data/green/train_image00312.jpg')
+    frame = cv2.imread('/home/evamo0508/Downloads/image.png')
```
Comment (kob51) [inline autovalet_lane_detection/utils/hsv_check.py]: did linux env variable for $HOME work?

File: autovalet_navigation/config/husky/global_costmap_params.yaml [hunk 11]
```diff
@@ -2,6 +2,12 @@ global_costmap:
   plugins:
       - {name: static_layer,   type: "costmap_2d::StaticLayer"}
       - {name: inflation,      type: "costmap_2d::InflationLayer"}
+      - {name: lane,           type: "costmap_2d::ObstacleLayer"}
+
+  lane:
+    observation_sources: fake_lane_sensor
+    fake_lane_sensor: {data_type: PointCloud2, sensor_frame: frontCamera_color_optical_frame, clearing: true, marking: true, topic: /lane/pointCloud}
+
   global_frame: map
```
Comment (kob51) [inline autovalet_navigation/config/husky/global_costmap_params.yaml:11]: will this need to be commented out when doing baseline planner testing? probably not right? just checking...

Comment (evamo0508) [reply autovalet_navigation/config/husky/global_costmap_params.yaml:11]: yea I'm not sure whether this should be pushed... this got into my branch while field testing. @SubramanianKrish what's your thoughts?

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py (hunk 3 above)
Comment (evamo0508) [reply autovalet_lane_detection/scripts/autovalet_lane_detector.py]: why? I thought `bb` was really common for bounding boxes. or alternatively, I'll make it `bbox`?

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py (hunk 5 above)
Comment (evamo0508) [reply autovalet_lane_detection/scripts/autovalet_lane_detector.py]: ok yea it's the default value when there are no lines detected. I'll comment more.

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py (hunk 6 above)
Comment (evamo0508) [reply autovalet_lane_detection/scripts/autovalet_lane_detector.py]: the Y direction is the opposite of what we usually think, and hence reversing one of the two to generate positive slope value. Agree it's a bit sloppy, any suggestions?

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py (hunk 3 above)
Comment (kob51) [reply autovalet_lane_detection/scripts/autovalet_lane_detector.py]: yea bbox works. i just keep thinking like that guy who gave the software lecture last year: reduce cognitive load in variable names

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py (hunk 6 above)
Comment (kob51) [reply autovalet_lane_detection/scripts/autovalet_lane_detector.py]: i think htat's fine just make a note of it 
something line `sign of slope is flipped here b/c origin is at top-left corner of image`

File: autovalet_husky/rviz/config.rviz (hunk 1 above)
Comment (evamo0508) [reply autovalet_husky/rviz/config.rviz]: removed

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py (hunk 2 above)
Comment (evamo0508) [reply autovalet_lane_detection/scripts/autovalet_lane_detector.py]: updated. see if that's enough.

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py (hunk 3 above)
Comment (evamo0508) [reply autovalet_lane_detection/scripts/autovalet_lane_detector.py]: updated. I understand, `bbox` would be a compromise between your longer version and my too short version.

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py (hunk 5 above)
Comment (evamo0508) [reply autovalet_lane_detection/scripts/autovalet_lane_detector.py]: comment a bit. see if that's enough.

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py (hunk 6 above)
Comment (evamo0508) [reply autovalet_lane_detection/scripts/autovalet_lane_detector.py]: updated.

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py (hunk 7 above)
Comment (evamo0508) [reply autovalet_lane_detection/scripts/autovalet_lane_detector.py:136]: inserted some comments above.

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py (hunk 8 above)
Comment (evamo0508) [reply autovalet_lane_detection/scripts/autovalet_lane_detector.py:158]: done.

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py (hunk 9 above)
Comment (evamo0508) [reply autovalet_lane_detection/scripts/autovalet_lane_detector.py:167]: put some comments right at the start of the function.

File: autovalet_lane_detection/utils/hsv_check.py (hunk 10 above)
Comment (evamo0508) [reply autovalet_lane_detection/utils/hsv_check.py]: this is just for testing, we won't have to run this piece of code eventually, but yea, I updated with the usage of `os.environ.get('HOME')`

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py (hunk 7 above)
Comment (kob51) [reply autovalet_lane_detection/scripts/autovalet_lane_detector.py:136]: where's s in the math? also just put in the def of K? I'm dumb sry.

here i did it for you
`K = [fx, 0, cx; 0, fy, cy; 0, 0, 0]` 

is it all zeros in the bottom row?

Comment (SubramanianKrish) [reply autovalet_lane_detection/scripts/autovalet_lane_detector.py:136]: s would be in row 1 column 2 like this: 'K = [fx, s, cx;  0, fy, cy;  0, 0, 0]'. We assume s is zero in the math

File: autovalet_navigation/config/husky/global_costmap_params.yaml (hunk 11 above)
Comment (SubramanianKrish) [reply autovalet_navigation/config/husky/global_costmap_params.yaml:11]: I think we can push it.. Anyways, it's not registering right now as an obstacle on the map. So no harm no foul. Push it and we'll debug using it!

**Review by SubramanianKrish — DISMISSED** [review]

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py (hunk 7 above)
Comment (evamo0508) [reply autovalet_lane_detection/scripts/autovalet_lane_detector.py:136]: what??? nooo.... s is not zero, we substitute it with z

File: autovalet_lane_detection/scripts/autovalet_lane_detector.py [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+        # ref: Subbu's code in goal_generation/scripts/transfer_lane.py
+        #  |u|    |x|
+        # s|v| = K|y|, where x,y,z is in rgb cam's frame.
+        #  |1|    |z|, x point to the right, y points downward, z points forward
+
+        u, v = coordinates[:, 0], coordinates[:, 1]
+        x    = (u - self.cx) / self.fx
+        y    = (v - self.cy) / self.fy
+
+        z = depth[v, u] / 1000.0;
```
Comment (kob51) [inline autovalet_lane_detection/scripts/autovalet_lane_detector.py]: units?

Comment (evamo0508) [reply autovalet_lane_detection/scripts/autovalet_lane_detector.py]: done

**Review by kob51 — APPROVED** [review]

**Review by SubramanianKrish — APPROVED** [review]

**Pull request merged**